model_compression_toolkit/core/pytorch/data_util.py

```diff
@@ -0,0 +1,161 @@
+from typing import Generator, Callable, Sequence, Any
+
+from torch.utils.data import IterableDataset, Dataset, DataLoader, default_collate
+
+from model_compression_toolkit.core.pytorch.utils import to_torch_tensor
+
+
+def flat_gen_fn(data_gen_fn: Callable[[], Generator]):
+    """
+    Convert data generator with arbitrary batch size to a flat (sample by sample) data generator.
+
+    Args:
+        data_gen_fn: input data generator factory. Generator is expected to yield lists of tensors.
+
+    Returns:
+        A factory for a flattened data generator.
+    """
+    def gen():
+        for inputs_batch in data_gen_fn():
+            for sample in zip(*inputs_batch):
+                yield to_torch_tensor(list(sample))
+    return gen
+
+
+class IterableDatasetFromGenerator(IterableDataset):
+    """
+    PyTorch iterable dataset built from a data generator factory.
+    Each iteration over the dataset corresponds to one pass over a fresh instance of a data generator.
+    Therefore, if the data generator factory creates data generator instances that yield different samples,
+    this behavior is preserved.
+    """
+
+    def __init__(self, data_gen_fn: Callable[[], Generator]):
+        """
+        Args:
+            data_gen_fn: a factory for data generator that yields lists of tensors.
+        """
+        # validate one batch
+        test_batch = next(data_gen_fn())
+        if not isinstance(test_batch, list):
+            raise TypeError(f'Data generator is expected to yield a list of tensors, got {type(test_batch)}')
+        self.orig_batch_size = test_batch[0].shape[0]
+
+        self._size = None
+        self._gen_fn = flat_gen_fn(data_gen_fn)
+
+    def __iter__(self):
+        """ Return an iterator for the dataset. """
+        return self._gen_fn()
+
+    def __len__(self):
+        """ Get the length of the dataset. """
+        if self._size is None:
+            self._size = sum(1 for _ in self)
+        return self._size
+
+
+class FixedDatasetFromGenerator(Dataset):
+    """
+    Dataset containing a fixed number of samples (i.e. same samples are yielded in each epoch), retrieved from a
+    data generator.
+    Note that the samples are stored in memory.
+
+    Attributes:
+        orig_batch_size: the batch size of the input data generator (retrieved from the first batch).
+    """
+    def __init__(self, data_gen_fn: Callable[[], Generator], n_samples: int = None):
+        """
+        Args:
+            data_gen_fn: data generator factory.
+            n_samples: target size of the dataset. If None, use all samples yielded by the data generator in one pass.
+        """
+        test_batch = next(data_gen_fn())
+        if not isinstance(test_batch, list):
+            raise TypeError(f'Data generator is expected to yield a list of tensors, got {type(test_batch)}')
+        self.orig_batch_size = test_batch[0].shape[0]
+
+        samples = []
+        for batch in data_gen_fn():
```

Comment:
I can't locate this in the changes so I'll just ask - is this, or any of the other dataset generators defined here, run shuffle on the dataset? meaning, are the batches returned with different samples in each epoch or with the same samples in each batch?

Idealy, for GPTQ, we would want shuffling in the dataset

Reply:
For fixed dataset shuffling can be defined in dataloader as it's a regular torch Dataset. 
For iterable dataset shuffling is problematic and torch doesn't allow it, so the provided data generator should take care of that. This is exactly the same as today. Iterable dataset follows the data generator precisely, I think its mentioned in the documentation